pkg/doc/presexch/definition.go

```diff
@@ -663,23 +663,79 @@ func createNewCredential(constraints *Constraints, src, limitedCred []byte,
 		return verifiable.ParseUnverifiedCredential(limitedCred, opts...)
 	}
 
+	limitedCred, err := enhanceRevealDoc(explicitPaths, limitedCred, src)
+	if err != nil {
+		return nil, err
+	}
+
+	var doc map[string]interface{}
+	if err := json.Unmarshal(limitedCred, &doc); err != nil {
+		return nil, err
+	}
+
+	return credential.GenerateBBSSelectiveDisclosure(doc, []byte(uuid.New().String()), opts...)
+}
+
+func enhanceRevealDoc(explicitPaths map[string]bool, limitedCred, vcBytes []byte) ([]byte, error) {
 	var err error
 
-	for _, path := range explicitPaths {
-		limitedCred, err = sjson.SetBytes(limitedCred, path, true)
+	limitedCred, err = sjson.SetBytes(limitedCred, "@explicit", true)
+	if err != nil {
+		return nil, err
+	}
+
+	for path, fromConstraint := range explicitPaths {
+		limitedCred, err = enhanceRevealField(path, limitedCred, vcBytes)
 		if err != nil {
 			return nil, err
 		}
+
+		if !fromConstraint {
+			continue
+		}
+
+		pathParts := strings.Split(path, ".")
+		combinedPath := ""
+
+		for i := 0; i < len(pathParts)-1; i++ {
+			if i == 0 {
+				combinedPath = pathParts[0]
+			} else {
+				combinedPath += "." + pathParts[i]
+			}
+
+			if _, ok := explicitPaths[combinedPath]; !ok {
+				explicitPaths[combinedPath] = false
+			}
+		}
 	}
 
-	var doc map[string]interface{}
-	if err := json.Unmarshal(limitedCred, &doc); err != nil {
+	return limitedCred, nil
+}
+
+func enhanceRevealField(path string, limitedCred, vcBytes []byte) ([]byte, error) {
+	var err error
+
+	limitedCred, err = sjson.SetBytes(limitedCred, path+".@explicit", true)
+	if err != nil {
 		return nil, err
 	}
 
-	doc["@explicit"] = true
+	for _, cf := range [...]string{"type", "@context"} {
+		specialFieldPath := path + "." + cf
 
-	return credential.GenerateBBSSelectiveDisclosure(doc, []byte(uuid.New().String()), opts...)
+		specialFieldValue := gjson.GetBytes(vcBytes, specialFieldPath)
+		if specialFieldValue.Type == gjson.Null {
+			continue
+		}
+
+		limitedCred, err = sjson.SetBytes(limitedCred, specialFieldPath, specialFieldValue.Value())
+		if err != nil {
+			return nil, err
```

Comment:
can these err types be wrapped?

Reply:
I just followed the error handling pattern in this source file...